src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,421 @@
+            SetMessageContent(request, message);
+            StartSend(client, message);
+        }
+
+        public void StartClientStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+        }
+
+        public void StartServerStreaming(System.Net.Http.HttpClient client, TRequest request)
+        {
+            var message = CreateHttpRequestMessage();
+            SetMessageContent(request, message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContextClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void StartDuplexStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContextClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void Dispose()
+        {
+            if (!Disposed)
+            {
+                Disposed = true;
+
+                _callCts.Cancel();
+                _callCts.Dispose();
+                _ctsRegistration?.Dispose();
+                _writerCtsRegistration?.Dispose();
+                _deadlineTimer?.Dispose();
+                HttpResponse?.Dispose();
+                ClientStreamReader?.Dispose();
+                ClientStreamWriter?.Dispose();
+            }
+        }
+
+        public void EnsureNotDisposed()
+        {
+            if (Disposed)
+            {
+                throw new ObjectDisposedException(nameof(GrpcCall<TRequest, TResponse>));
+            }
+        }
+
+        public void EnsureHeadersValid()
+        {
+            if (_headerValidationError != null)
+            {
+                throw new InvalidOperationException(_headerValidationError);
+            }
+        }
+
+        public Exception CreateCanceledStatusException()
+        {
+            if (_headerValidationError != null)
+            {
+                return new InvalidOperationException(_headerValidationError);
+            }
+
+            var statusCode = DeadlineReached ? StatusCode.DeadlineExceeded : StatusCode.Cancelled;
+            return new RpcException(new Status(statusCode, string.Empty));
+        }
+
+        public void FinishResponse(HttpResponseMessage httpResponseMessage)
+        {
+            if (ResponseFinished)
+            {
+                return;
+            }
+
+            ResponseFinished = true;
+
+            // Clean up call resources once this call is finished
+            // Call may not be explicitly disposed when used with unary methods
+            // e.g. var reply = await client.SayHelloAsync(new HelloRequest());
+            Dispose();
+
+            HttpResponse = httpResponseMessage;
```

Comment:
This is odd, you've called dispose at this point but you're modifying the state of `this` after disposing. Is there a reason you need to set the field or can you just use the local for `GetStatusCore`?

Reply:
Ah. Setting this isn't required anymore because it happens when awaiting the send task.